actors/market/src/lib.rs

```diff
@@ -1025,6 +1008,122 @@ impl Actor {
+        let mut total_slashed = TokenAmount::zero();
+
+        rt.transaction(|st: &mut State, rt| {
+            let mut new_deal_states: Vec<(DealID, DealState)> = Vec::new();
+            for deal_id in params.deal_ids {
+                let deal_proposal = match st.get_proposal(rt.store(), deal_id) {
+                    Ok(prop) => prop,
+                    Err(_) => {
+                        batch_gen.add_fail(ExitCode::USR_NOT_FOUND); // hide potential EX_DEAL_EXPIRED as deal may have terminated
+                        continue;
+                    }
+                };
+                let dcid = match rt_deal_cid(rt, &deal_proposal) {
+                    Ok(cid) => cid,
+                    Err(e) => {
+                        batch_gen.add_fail(e.exit_code());
+                        continue;
+                    }
+                };
+
+                let loaded_deal = match st.get_active_deal_or_process_timeout(
+                    rt.store(),
+                    curr_epoch,
+                    deal_id,
+                    &deal_proposal,
+                    &dcid,
+                ) {
+                    Ok(res) => res,
+                    Err(e) => {
+                        batch_gen.add_fail(e.exit_code());
+                        continue;
+                    }
+                };
+
+                let mut deal_state = match loaded_deal {
+                    LoadDealState::TooEarly => {
+                        // deal is not active, we process it as a zero-payment no-op
+                        settlements.push(DealSettlementSummary {
+                            completed: false,
+                            payment: TokenAmount::zero(),
+                        });
+                        continue;
+                    }
+                    LoadDealState::ProposalExpired(penalty) => {
+                        // deal proposal was not activated in time
+                        total_slashed += penalty;
+                        batch_gen.add_fail(ExitCode::USR_NOT_FOUND);
+                        continue;
+                    }
+                    LoadDealState::Loaded(deal_state) => deal_state,
+                };
+
+                let (slash_amount, payment_amount, remove_deal) = match st.process_deal_update(
+                    rt.store(),
+                    &deal_state,
+                    &deal_proposal,
+                    &dcid,
+                    curr_epoch,
+                ) {
+                    Ok(res) => res,
+                    Err(e) => {
+                        batch_gen.add_fail(e.exit_code());
+                        continue;
+                    }
+                };
+
+                // TODO: remove this case when it becomes impossible for process_deal_update to encounter slashed deals
+                // https://github.com/filecoin-project/builtin-actors/issues/1388
+                if !slash_amount.is_zero() {
```

Comment:
Using slash amount to detect termination is a little fragile. Better would be to inspect the state slash_epoch _before_ calling process_deal_update, and abort then. Then you can ignore the slash_amount return value.